fineract-provider/src/main/java/org/apache/fineract/cob/loan/LoanCOBWorkerConfiguration.java

```diff
@@ -0,0 +1,122 @@
+ * under the License.
+ */
+package org.apache.fineract.cob.loan;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.fineract.cob.COBPropertyService;
+import org.apache.fineract.portfolio.loanaccount.domain.Loan;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanRepository;
+import org.springframework.batch.core.Job;
+import org.springframework.batch.core.Step;
+import org.springframework.batch.core.configuration.annotation.JobBuilderFactory;
+import org.springframework.batch.core.configuration.annotation.StepScope;
+import org.springframework.batch.core.launch.support.RunIdIncrementer;
+import org.springframework.batch.integration.partition.RemotePartitioningWorkerStepBuilderFactory;
+import org.springframework.batch.item.ItemProcessor;
+import org.springframework.batch.item.ItemReader;
+import org.springframework.batch.item.data.RepositoryItemWriter;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.integration.channel.QueueChannel;
+import org.springframework.integration.dsl.IntegrationFlow;
+import org.springframework.integration.dsl.IntegrationFlows;
+import org.springframework.integration.event.core.MessagingEvent;
+import org.springframework.integration.event.inbound.ApplicationEventListeningMessageProducer;
+
+@Configuration
+@ConditionalOnProperty(value = "fineract.mode.batch-worker-enabled", havingValue = "true")
+public class LoanCOBWorkerConfiguration {
+
+    @Autowired
+    private JobBuilderFactory jobBuilderFactory;
+    @Autowired
+    private RemotePartitioningWorkerStepBuilderFactory stepBuilderFactory;
+    @Autowired
+    private COBPropertyService cobPropertyService;
+    @Autowired
+    private LoanRepository loanRepository;
+
+    private static final String JOB_NAME = "LOAN_COB";
+
+    @Bean(name = "Loan COB worker")
+    public Step loanCOBWorkerStep() {
+        return stepBuilderFactory.get("Loan COB worker").inputChannel(inboundRequests())
+                .<Loan, Loan>chunk(cobPropertyService.getChunkSize(JOB_NAME)).reader(itemReader(null)).processor(itemProcessor())
+                .writer(itemWriter()).build();
+    }
+
+    @Bean
+    public Job loanCOBWorkerJob() {
+        return jobBuilderFactory.get("Loan COB worker").start(loanCOBWorkerStep()).incrementer(new RunIdIncrementer()).build();
+    }
+
+    @Bean
+    @StepScope
+    public ItemReader<Loan> itemReader(@Value("#{stepExecutionContext['data']}") List<Integer> data) {
+        List<Integer> remainingData = new ArrayList<>(data);
+        return () -> {
+            if (remainingData.size() > 0) {
+                return loanRepository.findById(remainingData.remove(0).longValue()).orElse(null);
```

Comment:
This seems to be an overkill to issue a single statement for every single ID, don't you think?
We shall rather grab a chunk size of Loans at once.